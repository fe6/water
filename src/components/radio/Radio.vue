<template>
  <div
    class="w-radio"
    :class="radioList"
    @click="clickFn"
    @mouseover="hoverStatus = true"
    @mouseout="hoverStatus = false"
  >
    <div
      class="w-radio-status"
      :class="{
        [`${preName}status-on`]: status,
        [`${preName}status-disabled`]: disabled,
        [`${preName}status-loading`]: loading,
      }"
      :style="borderColorStyle"
    >
      <input
        class="w-radio-input"
        type="radio"
        name="inputName"
        :checked="status"
        :value="label"
        aria-label="radio"
      />
      <i
        class="w-radio-inner"
        :class="{
          [`${preName}inner-on`]: status,
          [`${preName}inner-disabled`]: disabled,
        }"
        :style="bgColorStyle"
      ></i>
    </div>
    <div class="w-radio-content" :style="textColorStyle">
      <slot></slot>
    </div>
  </div>
</template>

<script lang="ts">
  import { Component, Model, Prop, Emit, Vue } from 'vue-property-decorator';

  export interface ChangeParamsEntity {
    ev: MouseEvent;
    value: string | number | boolean;
    label: string | number | boolean;
  }

  export interface ReturnParamsEntity extends ChangeParamsEntity {
    status: boolean;
  }

  @Component
  export default class Radio extends Vue {
    name = 'Radio';

    preName = 'w-radio-';

    hoverStatus = false;

    @Model('model', {
      type: [String, Number, Boolean],
      default: '',
    })
    readonly value!: string | number | boolean;

    @Prop([String, Number, Boolean]) private label!: string | number | boolean;

    @Prop(Boolean) private disabled!: boolean;

    @Prop(Boolean) private loading!: boolean;

    @Prop(String) private textColor!: string;

    @Prop(String) private hoverColor!: string;

    @Prop(String) private checkColor!: string;

    @Prop([String, Object, Array]) private className!: string | object | any[];

    @Prop({
      type: Function,
      default() {
        return new Promise((resolve) => {
          resolve();
        });
      },
    })
    private before!: Function;

    @Prop({
      type: Function,
      default: () => {},
    })
    private change?: Function;

    get borderColorStyle(): string {
      let colors = '';
      const colorKey = 'border-color: ';

      if (this.hoverStatus && this.hoverColor) {
        colors = `${colorKey}${this.hoverColor}`;
      }

      if (this.status && this.checkColor) {
        colors = `${colorKey}${this.checkColor}`;
      }

      return colors;
    }

    get textColorStyle(): string {
      let colors = '';

      if (this.textColor) {
        colors = `color: ${this.textColor}`;
      }

      return colors;
    }

    get bgColorStyle(): string {
      let colors = '';

      if (this.status && this.checkColor) {
        colors = `background-color: ${this.checkColor}`;
      }

      return colors;
    }

    get radioList(): any[] {
      return [
        {
          [`${this.preName}on`]: this.status,
          [`${this.preName}disabled`]: this.disabled,
        },
        this.className,
      ];
    }

    get status(): boolean {
      return this.label === this.value;
    }

    @Emit('model')
    clickFn(ev: MouseEvent): string | number | boolean {
      if (!this.disabled && !this.loading && !this.status) {
        const reParams: ChangeParamsEntity = {
          ev,
          value: this.label,
          label: this.label,
        };
        return (this.before as Function)(reParams).then(() => {
          this.emitChange(ev);
          return this.label;
        });
      }
      return this.value;
    }

    @Emit('change')
    emitChange(ev: MouseEvent): ReturnParamsEntity {
      const reParams: ReturnParamsEntity = {
        ev,
        value: this.label,
        label: this.label,
        status: this.status,
      };

      (this.change as Function)(reParams);

      return reParams;
    }
  }
</script>

<style lang="scss">
  @import './radio.scss';
</style>
