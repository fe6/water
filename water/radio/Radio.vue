<template>
  <div class="w-radio" :class="[`w-radio-${modeString}`, {
    [`${prefixCls}click`]: isClick,
    [`${prefixCls}on`]:status,
    [`${prefixCls}disabled`]: disabledStatus,
    [`${prefix}-radio`]: !!prefix,
    [`${prefix}-radio`]: !!prefix,
    [`${prefix}-radio-${modeString}`]: !!prefix,
    [`${prefix}-click`]: isClick && !!prefix,
    [`${prefix}-on`]:status && !!prefix,
    [`${prefix}-disabled`]: disabledStatus && !!prefix,
  }]" @click="clickFn($event)" @animationend="removeClickName">
    <div class="w-radio-status" :class="{
      [`${prefix}-radio-status`]: !!prefix,
      ['w-radio-status-disabled']: disabledStatus,
      [`${prefix}-radio-status-disabled`]: disabledStatus && !!prefix,
      ['w-radio-status-loading']: loadingStatus,
    }">
      <input class="w-radio-input" type="radio" :name="inputName" :value="key" :class="{
        [`${prefix}-radio-input`]: !!prefix,
      }">
      <i class="w-radio-inner" :class="{
        [`${prefix}-radio-inner`]: !!prefix,
        ['w-radio-inner-disabled']: disabledStatus,
        [`${prefix}-radio-inner-disabled`]: disabledStatus && !!prefix,
      }"></i>
    </div>
    <div class="w-radio-content" :class="{
      [`${prefix}-radio-content`]: !!prefix,
    }">
      <slot></slot>
    </div>
  </div>
</template>
<script>
const radioName = 'w-radio';
const prefixCls = `${radioName}-`;

export default {
  name: 'WRadio',
  data() {
    const { $parent, $vnode } = this;
    const { key = 0 } = $vnode.data;
    const {
      index = 0, $options, name, defaultChecked = true, prefix, disabled, loading,
    } = $parent;

    return {
      inputName: name || this.name || 'wradio',
      modeString: '',
      prefixCls,
      isClick: false,
      defaultChecked,
      disabledStatus: disabled || this.disabled,
      loadingStatus: loading,
      prefix,
      key,
      index,
      /* eslint-disable no-underscore-dangle */
      status: index === key && $options._componentTag === `${prefixCls}group`,
    };
  },
  props: {
    name: String,
    disabled: Boolean,
  },
  methods: {
    clickFn(evente) {
      const {
        loading,
        before,
      } = this.$parent;
      if (!this.disabledStatus && !loading && !this.status) {
        before().then(() => {
          this.changeStatus(evente);
        });
      }
    },
    changeStatus(evente) {
      const { key } = this.$vnode.data;
      this.isClick = true;
      this.$parent.index = key || 0;
      this.$parent.updateChild();
      this.$parent.changeFn(evente, key);
    },
    removeClickName() {
      this.isClick = false;
    },
    /*
     * 父组件更新子组件选中状态的方法
     */
    updateIndex() {
      const { index = 0 } = this.$parent;
      const { key = 0 } = this.$vnode.data;
      this.index = index;
      this.status = index === key;
    },
    setMode(value = 'horizontal') {
      this.modeString = value;
    },
    /*
     * 父组件更新子组件禁用状态的方法
     */
    setDisabledStatus(value) {
      this.disabledStatus = value || this.disabled;
    },
    setDisabledOneStatus(value) {
      this.disabledStatus = this.$parent.disabled || value;
    },
    setLoadingStatus(value) {
      this.loadingStatus = value && !this.status;
    },
  },
  watch: {
    disabled(value) {
      this.setDisabledOneStatus(value);
    },
  },
};
</script>
