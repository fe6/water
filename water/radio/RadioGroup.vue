<template>
  <div class="w-radio-group" :class="[`w-radio-group-${modeString}`, {
    [`${prefix}-radio-group`]: !!prefix,
  }]">
    <slot></slot>
  </div>
</template>
<script>
export default {
  name: 'wRadioGroup',
  data() {
    return {
      index: this.defaultChecked ? this.value || this.on : -1,
      modeString: '',
      disabledStatus: false,
    };
  },
  props: {
    name: String,
    on: {
      type: Number,
      default: 0,
    },
    prefix: String,
    disabled: Boolean,
    loading: Boolean,
    defaultChecked: {
      type: Boolean,
      default: true,
    },
    mode: {
      type: String,
      default: 'horizontal', // 默认 水平( horizontal ) & 垂直( vertical )
    },
    value: Number,
    // before: Function,
    before: {
      type: Function,
      default() {
        return new Promise((resolve, reject) => {
          resolve();
        });
      },
    },
    change: {
      type: Function,
      default: () => {},
    },
  },
  mounted() {
    // 设置模式, 水平 & 垂直
    this.setMode(this.mode);
    this.updateChild();
    this.setDisabled(this.disabled);
    this.setLoading(this.loading);
  },
  methods: {
    updateChild() {
      this.$children.forEach((child) => {
        child.updateIndex();
      });
    },
    changeFn(evente, key) {
      this.change(evente, key);
      this.$emit('input', key);
      this.$emit('change', evente, key);
    },
    setMode(value) {
      this.modeString = value;
      this.$children.forEach((child) => {
        child.setMode(value);
      });
    },
    setDisabled(value) {
      this.disabledStatus = value;
      this.$children.forEach((child) => {
        child.setDisabledStatus(value);
      });
    },
    setLoading(value) {
      this.loadingStatus = value;
      this.$children.forEach((child) => {
        child.setLoadingStatus(value);
      });
    },
  },
  watch: {
    on(val) {
      this.index = val;
      this.updateChild();
    },
    mode(val) {
      this.setMode(val);
    },
    disabled(val) {
      this.setDisabled(val);
    },
    loading(val) {
      this.setLoading(val);
    },
  },
};
</script>
