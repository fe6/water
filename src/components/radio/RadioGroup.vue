<template>
  <div class="w-radio-group" :class="className">
    <Radio
      v-for="(item, itemIndex) in options"
      :key="item[fieldNames.key] || itemIndex"
      :disabled="disabled || item[fieldNames.disabled]"
      :value="checkValue"
      :class="radioClassName"
      :label="item[fieldNames.label]"
      :before="before.bind(this, item)"
      :loading="item[fieldNames.loading]"
      :hover-color="hoverColor"
      :check-color="checkColor"
      @change="radioChange"
      >{{ item[fieldNames.value] }}</Radio
    >
  </div>
</template>

<script lang="ts">
  import { Component, Model, Prop, Emit, Vue } from 'vue-property-decorator';
  import Radio, { ReturnParamsEntity } from './Radio.vue';

  interface OptionsEntity {
    [propName: string]: any;
  }

  interface FieldNamesEntity {
    value: string;
    label: string;
    disabled: string;
    loading: string;
    key?: number | string;
  }

  @Component({
    components: {
      Radio,
    },
  })
  export default class RadioGroup extends Vue {
    name = 'RadioGroup';

    @Model('model', { type: [String, Number, Boolean], default: '' })
    readonly value!: string | number | boolean;

    @Prop({
      type: Array,
      default: () => [],
    })
    private options!: OptionsEntity[];

    @Prop({
      type: Object,
      default: (): FieldNamesEntity => ({
        value: 'value',
        label: 'label',
        key: 'key',
        loading: 'loading',
        disabled: 'disabled',
      }),
    })
    private fieldNames!: FieldNamesEntity;

    @Prop(Boolean) private disabled!: boolean;

    @Prop(String) private hoverColor!: string;

    @Prop(String) private checkColor!: string;

    @Prop([String, Object, Array]) private className!: string | object | any[];

    @Prop([String, Object, Array]) private radioClassName!:
      | string
      | object
      | any[];

    @Prop({
      type: Function,
      default() {
        return new Promise((resolve) => {
          resolve();
        });
      },
    })
    private before!: Function;

    @Prop({
      type: Function,
      default: () => {},
    })
    private change?: Function;

    get checkValue(): string | number | boolean {
      return this.value;
    }

    @Emit('change')
    radioChange(params: ReturnParamsEntity) {
      this.radioModel(params);
    }

    @Emit('model')
    radioModel(params: ReturnParamsEntity): string | number | boolean {
      (this.change as Function)(params);
      return params.value;
    }
  }
</script>

<style lang="scss">
  @import 'radiogroup.scss';
</style>
