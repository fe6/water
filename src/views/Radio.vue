<template>
  <div>
    {{base}}=
    {{base1}}=
    <p>
      <Radio
        hoverColor="#f00"
        checkColor="#0f0"
        v-model="base1"
        :label="true"
      >定制</Radio>
    </p>
    <p>
      <Radio
        disabled
      >123</Radio>
      <Radio
        disabled
        value="true"
        label="true"
      >123</Radio>
    </p>
    <p>
      <Radio
        v-model="base"
        v-for="opt in options"
        :key="opt.age"
        :label="opt.label"
        :before="beforeLoading.bind(this, opt)"
        :loading="opt.loading"
        :change="change"
        @change="change"
      >{{opt.value}}</Radio>
      <br>
      {{group}}
      <br>
      <RadioGroup
        :options="options"
        v-model="group"
        :before="beforeLoading"
        :change="change"
        @change="change"
      />
      <button @click="group = 'tv'">设置电视</button>
    </p>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import Radio from '@/components/radio/Radio.vue';
import RadioGroup from '@/components/radio/RadioGroup.vue';

interface GroupOptsEntity {
  value: string;
  age: string;
  label: string;
  disabled?: boolean;
  more?: boolean;
}

@Component({
  components: {
    Radio,
    RadioGroup,
  },
})
export default class RadioView extends Vue {
  base: string = 'apple';

  base1: boolean = false;

  options: any[] = [
    {
      value: '苹果',
      label: 'apple',
      loading: false,
      age: 18,
    },
    {
      value: '相机',
      label: 'camera',
      loading: false,
      age: 1,
    },
    {
      value: '电视',
      label: 'tv',
      loading: false,
      age: 90,
    },
  ];

  loading: boolean = false;

  group: string = 'apple';

  /* eslint-disable class-methods-use-this */
  change(params: any) {
    console.log(params, 'chnage');
  }

  /* eslint-disable class-methods-use-this */
  beforeLoading(opt: any) {
    /* eslint-disable no-alert */
    opt.loading = true;
    return new Promise((resolve) => {
      setTimeout(() => {
        opt.loading = false;
        if (window.confirm('你确定选择吗？')) {
          resolve();
        }
      }, 2000);
    });
  }
}
</script>
