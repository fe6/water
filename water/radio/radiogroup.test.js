import { shallow } from 'vue-test-utils';
import Radio from './Radio';
import RadioGroup from './RadioGroup';

describe('RadioGroup.vue', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallow(RadioGroup, {
      slots: {
        default: [Radio],
      },
    });
  });

  it('检测快照是否一样。', (done) => {
    wrapper.vm.$nextTick(() => {
      try {
        expect(wrapper.vm.$el).toMatchSnapshot();
        done();
      } catch (err) {
        done.fail(err);
      }
    });
  });

  it('默认第一个选中', (done) => {
    expect(wrapper.vm.$children[0].status).toBeTruthy();
    done();
  });

  it('索引变化是否有样式更新', (done) => {
    expect(wrapper.vm.$children[0].status).toBeTruthy();
    // 是否触发 watch
    const updateChildFn = jest.fn();
    wrapper.vm.updateChild = updateChildFn;
    wrapper.vm.on = 2;
    wrapper.update();
    expect(wrapper.vm.index).toBe(2);
    expect(wrapper.vm.updateChild).toBeCalled();
    // 是否触发 watch mode
    const setModeFn = jest.fn();
    wrapper.vm.setMode = setModeFn;
    wrapper.vm.mode = 'vertical';
    wrapper.update();
    expect(wrapper.vm.mode).toBe('vertical');
    expect(wrapper.vm.setMode).toBeCalled();
    // 是否触发 watch disabled
    const setDisabledFn = jest.fn();
    wrapper.vm.setDisabled = setDisabledFn;
    wrapper.vm.disabled = true;
    wrapper.update();
    expect(wrapper.vm.disabled).toBeTruthy();
    expect(wrapper.vm.setDisabled).toBeCalled();
    // 是否触发 watch loading
    const setLoadingFn = jest.fn();
    wrapper.vm.setLoading = setLoadingFn;
    const setLoadingStatusFn = jest.fn();
    wrapper.vm.$children[0].setLoadingStatus = setLoadingStatusFn;
    wrapper.vm.loading = true;
    wrapper.update();
    expect(wrapper.vm.loading).toBeTruthy();
    expect(wrapper.vm.setLoading).toBeCalled();
    setTimeout(() => {
      expect(wrapper.vm.$children[0].setLoadingStatus).toBeCalled();
    }, 0);
    done();
  });
});
