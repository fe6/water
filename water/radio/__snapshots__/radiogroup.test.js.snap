// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup.vue 检测快照是否一样。 1`] = `
<div
  class="w-radio-group w-radio-group-horizontal"
>
  <div
    class="w-radio w-radio-horizontal w-radio-on"
  >
    <div
      class="w-radio-status"
    >
      <input
        class="w-radio-input"
        name="wradio"
        type="radio"
        value="0"
      />
       
      <i
        class="w-radio-inner"
      />
    </div>
     
    <div
      class="w-radio-content"
    />
  </div>
</div>
`;
